### PR TITLE
Added error when a customization is not found

### LINKED ISSUE #22110: Multistore - BO - Order page - An unexpected error occurred when adding a product with customization
#### Describe the bug

When multishop is enabled, a new shop is created => Customized products are imported in the second shop but without customization as reported [here](https://github.com/PrestaShop/PrestaShop/issues/18919)
This issue will occur a problem in the Add new Order page
image

#### Steps to Reproduce

1. Enable Multishop context 
2. Create a new shop with importing all data
3. `Shop2` is selected Go to BO > Orders > Add new order page 
4. Choose a customer
5. Search for the product `demo_14` "Customizable mug"
6. Click Add
7. See error
image

PS1768
image

#### Additional information

* PrestaShop version: 1770stablebuild1
* PHP version: 7.2

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | MultiStore - Added error when a customization is not found
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #22110
| How to test?  | Cf. #22110

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/22175)

### CODE REVIEW

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/CartController.php [hunk 1]
```diff
@@ -587,6 +588,8 @@ private function getErrorMessages(Exception $e)
                     ['%limit%' => CustomizationSettings::MAX_TEXT_LENGTH]
                 ),
             ],
+            ProductCustomizationNotFoundException::class => $this->trans(
```
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Sell/Order/CartController.php:591]: No wording here...

File: tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php [hunk 2]
```diff
@@ -196,15 +197,21 @@ public function updateProductQuantityInCart(int $quantity, string $productName,
     }
 
     /**
-     * @When I add :quantity customized products with reference :productReference to the cart :reference
+     * @When I add :quantity customized products with its customization with reference :productReference to the cart :reference
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php]:
```suggestion
     * @When I add :quantity customized products with reference :productReference  with all its customizations to the cart :reference
```

File: tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php [hunk 3]
```diff
@@ -196,15 +197,21 @@ public function updateProductQuantityInCart(int $quantity, string $productName,
...
-    {
+    public function addCustomizedProductToCartsWithCustomization(
+        int $quantity,
+        string $productReference,
+        string $reference
+    ) {
         $productId = (int) Product::getIdByReference($productReference);
         $product = new Product($productId);
-        $customizationFields = $product->getCustomizationFieldIds();
+        $customizationsFields = $product->getCustomizationFields();
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php]: Why did you change this? The `$customizationsFields` is only used once to check if it's empty and then you use `$product->getCustomizationFieldIds()` like before?

File: tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php [hunk 4]
```diff
@@ -231,6 +238,45 @@ public function addCustomizedProductToCarts(int $quantity, $productReference, $r
         );
     }
 
+    /**
+     * @When I add :quantity customized products without its customization with reference :productReference to the cart :reference
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php]:
```suggestion
     * @When I add :quantity customized products with reference :productReference without its customization to the cart :reference
```

File: tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php [hunk 5]
```diff
@@ -231,6 +238,45 @@ public function addCustomizedProductToCarts(int $quantity, $productReference, $r
...
+        }
+        $cartId = (int) SharedStorage::getStorage()->get($reference);
+
+        try {
+            $this->getCommandBus()->handle(
+                new UpdateProductQuantityInCartCommand(
+                    $cartId,
+                    $productId,
+                    $quantity,
+                    null,
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php:255]: Maybe you should use empty `array` rather than `null` ?

**Review by jolelievre — CHANGES_REQUESTED** [review]
Some minor improvement However the missing wording related to the new exception is mandatory

**Comment by jolelievre** [comment]
However I don't understand the relation with the Multi Store? 🤷‍♂️

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/CartController.php (hunk 1 above)
Comment (Progi1984) [reply src/PrestaShopBundle/Controller/Admin/Sell/Order/CartController.php:591]: Fixed

File: tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php (hunk 3 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php]: Fixed

File: tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php (hunk 5 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php:255]: No Type is `int|null`

File: tests/Integration/Behaviour/Features/Scenario/Order/order_to_cart.feature [hunk 6]
```diff
@@ -13,11 +13,20 @@ Feature: Check order to cart data copy
...
     # Additional checks should be added to validate the cart is correctly duplicated (even customization address)
+
+  @order-to-cart
+  Scenario: Add a customized product without its customization
+    Given there is customer "customer_for_customization" with email "pub@prestashop.com"
+    And customer "customer_for_customization" has address in "US" country
+    And the module "dummy_payment" is installed
+    When I create an empty cart "dummy_custom_cart" for customer "customer_for_customization"
+    And add 1 customized product with reference "demo_14" without all its customizations to the cart "dummy_custom_cart"
+    Then I should get an error that the product is customizable and the customization is not provided
```
Comment (matthieu-rolland) [inline tests/Integration/Behaviour/Features/Scenario/Order/order_to_cart.feature:32]: @Progi1984  Did you manage to reproduce the issue with Behat without being in multishop mode ?

**Review by matthieu-rolland — COMMENTED** [review]
just one question (otherwise PR seems ok)

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/CartController.php (hunk 1 above)
Comment (Julievrz) [reply src/PrestaShopBundle/Controller/Admin/Sell/Order/CartController.php:591]: Hi @Progi1984 I updated the wording here (https://github.com/PrestaShop/PrestaShop/issues/22110#issuecomment-736640982) 

"Product customization could not be found. Go to Catalog > Products to customize the product." in Admin.Catalog.Notification

Comment (Progi1984) [reply src/PrestaShopBundle/Controller/Admin/Sell/Order/CartController.php:591]: Fixed @Julievrz

File: tests/Integration/Behaviour/Features/Scenario/Order/order_to_cart.feature (hunk 6 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Order/order_to_cart.feature:32]: With code yes.
In real, you must use the multishop (duplicate) which is broken.

**Review by jolelievre — DISMISSED** [review]

**Review by matthieu-rolland — DISMISSED** [review]

**Review by kpodemski — DISMISSED** [review]

**Comment by SD1982** [comment]
LGTM Thanks @Progi1984 !!

File: src/Adapter/Cart/CommandHandler/UpdateProductQuantityInCartHandler.php [hunk 7]
```diff
@@ -205,12 +206,15 @@ private function assertProductIsInStock(Product $product, UpdateProductQuantityI
...
      *
-     * @throws ProductException
+     * @throws ProductCustomizationNotFoundException
      */
     private function assertProductCustomization(Product $product, UpdateProductQuantityInCartCommand $command)
     {
         if (null === $command->getCustomizationId() && !$product->hasAllRequiredCustomizableFields()) {
-            throw new ProductException(sprintf('Missing customization for product with id "%s"', $product->id));
+            throw new ProductCustomizationNotFoundException(sprintf(
+                'Missing customization for product with id "%s"',
```
Comment (PierreRambaud) [inline src/Adapter/Cart/CommandHandler/UpdateProductQuantityInCartHandler.php:215]:
```suggestion
                'Missing customization for product with id "%d"',
```

File: tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php [hunk 8]
```diff
@@ -209,39 +210,63 @@ public function updateProductQuantityInCart(int $quantity, string $productName,
     }
 
     /**
-     * @When I add :quantity customized products with reference :productReference to the cart :reference
+     * @When /^(?:|I )add (\d+) customized product(?:|s) with reference "(.+)" (with|without) all its customizations to the cart "(.+)"$/
```
Comment (PierreRambaud) [inline tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php]:
```suggestion
     * @When /^(?:I )?add (\d+) customized products? with reference "(.+)" with(?:out)? all its customizations to the cart "(.+)"$/
```

**Review by PierreRambaud — DISMISSED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Review by atomiix — APPROVED** [review]

**Comment by atomiix** [comment]
Thanks @Progi1984
